File: internal/harness/registry.go
package harness

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"

	"github.com/danieljustus/symaira-corekit/exitcodes"
)

// Name identifies one harness known to symbrain.
type Name string

// The harness names known to symbrain.  Capability-only entries are
// included here too, so downstream registries cannot silently drift.
const (
	Claude        Name = "claude"
	ClaudeDesktop Name = "claude-desktop"
	Cursor        Name = "cursor"
	Opencode      Name = "opencode"
	Codex         Name = "codex"
	Gemini        Name = "gemini"
	Agents        Name = "agents"
	Hermes        Name = "hermes"
	Antigravity   Name = "antigravity"
	OpenClaw      Name = "openclaw"
)

// InstructionAdapter identifies the instruction-file adapter used by a
// harness.  The value is an adapter kind, not another harness registry.
type InstructionAdapter string

const (
	InstructionAdapterNone   InstructionAdapter = ""
	InstructionAdapterAgents InstructionAdapter = "agents"
	InstructionAdapterClaude InstructionAdapter = "claude"
	InstructionAdapterCursor InstructionAdapter = "cursor"
	InstructionAdapterGemini InstructionAdapter = "gemini"
)

// SkillTarget identifies the in-process skills renderer target for a harness.
type SkillTarget string

const (
	SkillTargetNone        SkillTarget = ""
	SkillTargetOpenCode    SkillTarget = "opencode"
	SkillTargetClaude      SkillTarget = "claude"
	SkillTargetCodex       SkillTarget = "codex"
	SkillTargetHermes      SkillTarget = "hermes"
	SkillTargetAntigravity SkillTarget = "antigravity"
	SkillTargetOpenClaw    SkillTarget = "openclaw"
)

// Format is the on-disk serialization of a harness's MCP config file.
type Format string

const (
	// FormatJSON configs are edited structurally via encoding/json.
	FormatJSON Format = "json"
	// FormatTOML configs are edited structurally via github.com/BurntSushi/toml.
	FormatTOML Format = "toml"
)

// ServerName is the key symbrain registers its MCP entry under in every
// harness's server map (e.g. mcpServers.symbrain, [mcp_servers.symbrain]).
const ServerName = "symbrain"

// Harness describes one known AI harness: its optional MCP config, and
// the instruction and skills capabilities symbrain can use with it.
type Harness struct {
	// Name is the stable identifier used on the CLI (--harness <name>).
	Name Name
	// DisplayName is the human-readable name shown by inspection commands.
	DisplayName string
	// Format is the config file's serialization.
	Format Format
	// ServersKey is the top-level key/table that holds the map of
	// server-name -> entry (e.g. "mcpServers" for the JSON harnesses,
	// "mcp_servers" for codex's TOML).
	ServersKey string

	// ConfigPath resolves this harness's primary (global, user-level)
	// config file path.
	ConfigPath func() (string, error)

	// SupportsMCPInstall reports whether symbrain can install its MCP
	// server entry into this harness's configuration.
	SupportsMCPInstall bool
	// InstructionAdapter selects the instruction-file adapter, when one
	// exists. An empty value is an intentional unsupported capability.
	InstructionAdapter InstructionAdapter
	// SkillTarget selects the in-process skills renderer target, when one
	// exists. An empty value is an intentional unsupported capability.
	SkillTarget SkillTarget

	// SupportsProject reports whether this harness also has a
	// project-local config file in addition to its global one.
	SupportsProject bool
	// ProjectConfigPath resolves the project-local config path rooted at
	// dir. Only meaningful when SupportsProject is true.
	ProjectConfigPath func(dir string) string
}

// All is the single registry of every harness known to symbrain. Entries
// that do not support MCP installation still participate in instruction or
// skills derivation when they declare those capabilities.
var All = []Harness{
	{
		Name:               Claude,
		DisplayName:        "Claude Code",
		Format:             FormatJSON,
		SupportsMCPInstall: true,
		InstructionAdapter: InstructionAdapterClaude,
		SkillTarget:        SkillTargetClaude,
		ServersKey:         "mcpServers",
		ConfigPath:         homeJoin(".claude.json"),
		SupportsProject:    true,
		ProjectConfigPath: func(dir string) string {
			return filepath.Join(dir, ".mcp.json")
		},
	},
	{
		Name:               ClaudeDesktop,
		DisplayName:        "Claude Desktop",
		Format:             FormatJSON,
		SupportsMCPInstall: true,
		InstructionAdapter: InstructionAdapterNone,
		SkillTarget:        SkillTargetNone,
		ServersKey:         "mcpServers",
		ConfigPath:         claudeDesktopConfigPath,
	},
	{
		Name:               Cursor,
		DisplayName:        "Cursor",
		Format:             FormatJSON,
		SupportsMCPInstall: true,
		InstructionAdapter: InstructionAdapterCursor,
		SkillTarget:        SkillTargetNone,
		ServersKey:         "mcpServers",
		ConfigPath:         homeJoin(".cursor", "mcp.json"),
	},
	{
		Name:               Opencode,
		DisplayName:        "OpenCode",
		Format:             FormatJSON,
		SupportsMCPInstall: true,
		InstructionAdapter: InstructionAdapterNone,
		SkillTarget:        SkillTargetOpenCode,
		ServersKey:         "mcpServers",
		ConfigPath:         xdgConfigJoin("opencode", "config.json"),
	},
	{
		Name:               Codex,
		DisplayName:        "Codex CLI",
		Format:             FormatTOML,
		SupportsMCPInstall: true,
		InstructionAdapter: InstructionAdapterNone,
		SkillTarget:        SkillTargetCodex,
		ServersKey:         "mcp_servers",
		ConfigPath:         homeJoin(".codex", "config.toml"),
	},
	{
		Name:               Gemini,
		DisplayName:        "Gemini CLI",
		Format:             FormatJSON,
		SupportsMCPInstall: true,
		InstructionAdapter: InstructionAdapterGemini,
		SkillTarget:        SkillTargetNone,
		ServersKey:         "mcpServers",
		ConfigPath:         homeJoin(".gemini", "settings.json"),
	},
	{
		Name:               Agents,
		DisplayName:        "AGENTS.md",
		SupportsMCPInstall: false,
		InstructionAdapter: InstructionAdapterAgents,
		SkillTarget:        SkillTargetNone,
		ConfigPath:         unsupportedConfigPath,
	},
	{
		Name:               Hermes,
		DisplayName:        "Hermes",
		SupportsMCPInstall: false,
		InstructionAdapter: InstructionAdapterNone,
		SkillTarget:        SkillTargetHermes,
		ConfigPath:         unsupportedConfigPath,
	},
	{
		Name:               Antigravity,
		DisplayName:        "Antigravity",
		SupportsMCPInstall: false,
		InstructionAdapter: InstructionAdapterNone,
		SkillTarget:        SkillTargetAntigravity,
		ConfigPath:         unsupportedConfigPath,
	},
	{
		Name:               OpenClaw,
		DisplayName:        "OpenClaw",
		SupportsMCPInstall: false,
		InstructionAdapter: InstructionAdapterNone,
		SkillTarget:        SkillTargetOpenClaw,
		ConfigPath:         unsupportedConfigPath,
	},
}

// Lookup finds a registered harness by name. An unknown name is reported as
// a typed *exitcodes.CLIError (ExitNoInput) so callers can refuse cleanly
// instead of guessing at a config location.
func Lookup(name string) (Harness, error) {
	for _, h := range All {
		if string(h.Name) == name {
			return h, nil
		}
	}
	return Harness{}, exitcodes.Wrap(
		fmt.Errorf("unknown harness %q", name),
		exitcodes.ExitNoInput,
		exitcodes.KindValidation,
		"harness: want one of: "+strings.Join(Names(), ", "),
	)
}

// Names returns every registered harness name, in registry order.
func Names() []string {
	names := make([]string, len(All))
	for i, h := range All {
		names[i] = string(h.Name)
	}
	sort.Strings(names)
	return names
}

func unsupportedConfigPath() (string, error) {
	return "", fmt.Errorf("harness does not support MCP installation")
}

// homeJoin returns a ConfigPath resolver for $HOME/elem...
func homeJoin(elem ...string) func() (string, error) {
	return func() (string, error) {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("resolve home directory: %w", err)
		}
		parts := append([]string{home}, elem...)
		return filepath.Join(parts...), nil
	}
}

// xdgConfigJoin returns a ConfigPath resolver for
// $XDG_CONFIG_HOME/elem... (default $HOME/.config/elem...), matching the
// XDG Base Directory convention used across the Symaira tools.
func xdgConfigJoin(elem ...string) func() (string, error) {
	return func() (string, error) {
		if base := os.Getenv("XDG_CONFIG_HOME"); base != "" {
			parts := append([]string{base}, elem...)
			return filepath.Join(parts...), nil
		}
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("resolve home directory: %w", err)
		}
		parts := append([]string{home, ".config"}, elem...)
		return filepath.Join(parts...), nil
	}
}

// claudeDesktopConfigPath resolves Claude Desktop's config file, which
// lives in a different location per OS:
//
//   - macOS:   ~/Library/Application Support/Claude/claude_desktop_config.json
//   - Linux:   $XDG_CONFIG_HOME/Claude/claude_desktop_config.json
//     (default ~/.config/Claude/claude_desktop_config.json)
//   - Windows: %APPDATA%\Claude\claude_desktop_config.json
func claudeDesktopConfigPath() (string, error) {
	return resolveClaudeDesktopConfigPath(runtime.GOOS, os.Getenv)
}

// resolveClaudeDesktopConfigPath is the testable core of
// claudeDesktopConfigPath: goos and getenv are injected so tests can cover
// every platform branch regardless of the OS actually running the test.
func resolveClaudeDesktopConfigPath(goos string, getenv func(string) string) (string, error) {
	const filename = "claude_desktop_config.json"

	switch goos {
	case "darwin":
		home, err := userHomeDirFor(getenv)
		if err != nil {
			return "", err
		}
		return filepath.Join(home, "Library", "Application Support", "Claude", filename), nil
	case "windows":
		if appData := getenv("APPDATA"); appData != "" {
			return filepath.Join(appData, "Claude", filename), nil
		}
		home, err := userHomeDirFor(getenv)
		if err != nil {
			return "", err
		}
		return filepath.Join(home, "AppData", "Roaming", "Claude", filename), nil
	default: // linux and other XDG-style platforms
		if base := getenv("XDG_CONFIG_HOME"); base != "" {
			return filepath.Join(base, "Claude", filename), nil
		}
		home, err := userHomeDirFor(getenv)
		if err != nil {
			return "", err
		}
		return filepath.Join(home, ".config", "Claude", filename), nil
	}
}

// userHomeDirFor mirrors os.UserHomeDir's platform lookup but reads through
// the injected getenv, so resolveClaudeDesktopConfigPath is fully
// deterministic under test.
func userHomeDirFor(getenv func(string) string) (string, error) {
	env := "HOME"
	if runtime.GOOS == "windows" {
		env = "USERPROFILE"
	}
	if v := getenv(env); v != "" {
		return v, nil
	}
	// Fall back to the real resolver for the environment this test double
	// doesn't cover (e.g. darwin/linux tests reading $HOME, which getenv
	// already handles above in practice).
	return os.UserHomeDir()
}

File: rust/symbrain-harness/src/registry.rs
use std::fmt;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::{HarnessError, relative_path, resolve_path, trusted_root};

/// Canonical server key written into harness configurations.
pub const SERVER_NAME: &str = "symbrain";

/// Stable identifier for a known harness.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum HarnessName {
    /// Claude Code.
    Claude,
    /// Claude Desktop.
    ClaudeDesktop,
    /// Cursor.
    Cursor,
    /// `OpenCode`.
    Opencode,
    /// Codex CLI.
    Codex,
    /// Antigravity.
    Antigravity,
    /// Generic `AGENTS.md` instruction target.
    Agents,
    /// Hermes skill target.
    Hermes,
    #[serde(rename = "openclaw")]
    /// `OpenClaw` skill target.
    OpenClaw,
}

/// Backward-compatible alias for [`HarnessName`].
pub type Name = HarnessName;

impl HarnessName {
    #[must_use]
    /// Returns the stable command-line name.
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Claude => "claude",
            Self::ClaudeDesktop => "claude-desktop",
            Self::Cursor => "cursor",
            Self::Opencode => "opencode",
            Self::Codex => "codex",
            Self::Antigravity => "antigravity",
            Self::Agents => "agents",
            Self::Hermes => "hermes",
            Self::OpenClaw => "openclaw",
        }
    }
}

impl fmt::Display for HarnessName {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

/// Configuration serialization format used by a harness.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Format {
    /// Ordered JSON.
    Json,
    /// TOML.
    Toml,
    /// No MCP configuration file is supported.
    Unsupported,
}

impl fmt::Display for Format {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Self::Json => "json",
            Self::Toml => "toml",
            Self::Unsupported => "unsupported",
        })
    }
}

/// Instruction-file adapter associated with a harness.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InstructionAdapter {
    /// No instruction adapter.
    None,
    /// Generic `AGENTS.md` adapter.
    Agents,
    /// Claude instruction adapter.
    Claude,
    /// Cursor instruction adapter.
    Cursor,
    /// Antigravity instruction adapter.
    Antigravity,
}

/// Skills installation target associated with a harness.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SkillTarget {
    /// No skills target.
    None,
    /// `OpenCode` target.
    OpenCode,
    /// Claude target.
    Claude,
    /// Codex target.
    Codex,
    /// Hermes target.
    Hermes,
    /// Antigravity target.
    Antigravity,
    /// `OpenClaw` target.
    OpenClaw,
}

#[derive(Debug, Clone, Copy)]
pub(crate) enum PathKind {
    Home(&'static [&'static str]),
    Xdg(&'static [&'static str]),
    ClaudeDesktop,
    Unsupported,
}

#[derive(Debug, Clone)]
/// A resolved configuration path plus the trusted root and relative capability path used to access it.
pub struct ConfigLocation {
    /// User-visible absolute or environment-derived path.
    pub path: PathBuf,
    /// Directory treated as the capability boundary.
    pub trusted_root: PathBuf,
    /// Target path relative to [`Self::trusted_root`].
    pub relative_path: PathBuf,
}

/// Static capabilities and path rules for one harness.
#[derive(Debug, Clone, Copy)]
pub struct Harness {
    /// Stable name.
    pub name: HarnessName,
    /// Human-readable name.
    pub display_name: &'static str,
    /// Configuration format.
    pub format: Format,
    /// Root key containing MCP servers, when supported.
    pub servers_key: Option<&'static str>,
    path_kind: PathKind,
    /// Whether MCP install/uninstall is supported.
    pub supports_mcp_install: bool,
    /// Instruction adapter.
    pub instruction_adapter: InstructionAdapter,
    /// Skills target.
    pub skill_target: SkillTarget,
    /// Whether project-local MCP configuration is supported.
    pub supports_project: bool,
}

impl Harness {
    /// Resolves this harness's config path from the process environment.
    ///
    /// # Errors
    /// Returns an error for unsupported harnesses or when the home directory is unavailable.
    pub fn config_path(&self) -> Result<PathBuf, HarnessError> {
        let env = std::env::vars().collect::<Vec<_>>();
        self.config_path_for(std::env::consts::OS, &env)
    }

    /// Resolves a path using an injected target OS and environment.
    ///
    /// # Errors
    /// Returns an error for unsupported harnesses or when the required home directory is absent.
    pub fn config_path_for(
        &self,
        target_os: &str,
        env: &[(String, String)],
    ) -> Result<PathBuf, HarnessError> {
        resolve_path(self.path_kind, target_os, env)
    }

    /// Resolves the configuration as a trusted-root capability target.
    ///
    /// The returned relative path is the only path opened beneath the trusted
    /// root; parent components and the final file are opened without following
    /// symlinks.
    ///
    /// # Errors
    /// Returns an error for unsupported harnesses or when the required home directory is absent.
    pub fn config_location(&self) -> Result<ConfigLocation, HarnessError> {
        let env = std::env::vars().collect::<Vec<_>>();
        self.config_location_for(std::env::consts::OS, &env)
    }

    /// Resolves a capability target using an injected target OS and environment.
    ///
    /// # Errors
    /// Returns an error for unsupported harnesses or when the required home directory is absent.
    pub fn config_location_for(
        &self,
        target_os: &str,
        env: &[(String, String)],
    ) -> Result<ConfigLocation, HarnessError> {
        Ok(ConfigLocation {
            path: self.config_path_for(target_os, env)?,
            trusted_root: trusted_root(self.path_kind, target_os, env)?,
            relative_path: relative_path(self.path_kind, target_os),
        })
    }

    #[must_use]
    /// Resolves the project-local configuration path when supported.
    pub fn project_config_path(&self, project_dir: &Path) -> Option<PathBuf> {
        self.supports_project.then(|| project_dir.join(".mcp.json"))
    }

    #[must_use]
    /// Returns the stable harness name.
    pub const fn name(&self) -> HarnessName {
        self.name
    }

    #[must_use]
    /// Returns the human-readable harness name.
    pub const fn display_name(&self) -> &'static str {
        self.display_name
    }

    #[must_use]
    /// Returns the harness configuration format.
    pub const fn format(&self) -> Format {
        self.format
    }

    #[must_use]
    /// Returns the MCP server-map key when supported.
    pub const fn server_key(&self) -> Option<&'static str> {
        self.servers_key
    }

    #[must_use]
    /// Reports whether MCP install/uninstall is supported.
    pub const fn supports_mcp_install(&self) -> bool {
        self.supports_mcp_install
    }
}

const HOME_CLAUDE: &[&str] = &[".claude.json"];
const HOME_CURSOR: &[&str] = &[".cursor", "mcp.json"];
const HOME_CODEX: &[&str] = &[".codex", "config.toml"];
const HOME_ANTIGRAVITY: &[&str] = &[".gemini", "config", "mcp_config.json"];
const XDG_OPENCODE: &[&str] = &["opencode", "config.json"];

static REGISTRY: [Harness; 9] = [
    Harness {
        name: HarnessName::Claude,
        display_name: "Claude Code",
        format: Format::Json,
        servers_key: Some("mcpServers"),
        path_kind: PathKind::Home(HOME_CLAUDE),
        supports_mcp_install: true,
        instruction_adapter: InstructionAdapter::Claude,
        skill_target: SkillTarget::Claude,
        supports_project: true,
    },
    Harness {
        name: HarnessName::ClaudeDesktop,
        display_name: "Claude Desktop",
        format: Format::Json,
        servers_key: Some("mcpServers"),
        path_kind: PathKind::ClaudeDesktop,
        supports_mcp_install: true,
        instruction_adapter: InstructionAdapter::None,
        skill_target: SkillTarget::None,
        supports_project: false,
    },
    Harness {
        name: HarnessName::Cursor,
        display_name: "Cursor",
        format: Format::Json,
        servers_key: Some("mcpServers"),
        path_kind: PathKind::Home(HOME_CURSOR),
        supports_mcp_install: true,
        instruction_adapter: InstructionAdapter::Cursor,
        skill_target: SkillTarget::None,
        supports_project: false,
    },
    Harness {
        name: HarnessName::Opencode,
        display_name: "OpenCode",
        format: Format::Json,
        servers_key: Some("mcpServers"),
        path_kind: PathKind::Xdg(XDG_OPENCODE),
        supports_mcp_install: true,
        instruction_adapter: InstructionAdapter::None,
        skill_target: SkillTarget::OpenCode,
        supports_project: false,
    },
    Harness {
        name: HarnessName::Codex,
        display_name: "Codex CLI",
        format: Format::Toml,
        servers_key: Some("mcp_servers"),
        path_kind: PathKind::Home(HOME_CODEX),
        supports_mcp_install: true,
        instruction_adapter: InstructionAdapter::None,
        skill_target: SkillTarget::Codex,
        supports_project: false,
    },
    Harness {
        name: HarnessName::Antigravity,
        display_name: "Antigravity",
        format: Format::Json,
        servers_key: Some("mcpServers"),
        path_kind: PathKind::Home(HOME_ANTIGRAVITY),
        supports_mcp_install: true,
        instruction_adapter: InstructionAdapter::Antigravity,
        skill_target: SkillTarget::Antigravity,
        supports_project: false,
    },
    Harness {
        name: HarnessName::Agents,
        display_name: "AGENTS.md",
        format: Format::Unsupported,
        servers_key: None,
        path_kind: PathKind::Unsupported,
        supports_mcp_install: false,
        instruction_adapter: InstructionAdapter::Agents,
        skill_target: SkillTarget::None,
        supports_project: false,
    },
    Harness {
        name: HarnessName::Hermes,
        display_name: "Hermes",
        format: Format::Unsupported,
        servers_key: None,
        path_kind: PathKind::Unsupported,
        supports_mcp_install: false,
        instruction_adapter: InstructionAdapter::None,
        skill_target: SkillTarget::Hermes,
        supports_project: false,
    },
    Harness {
        name: HarnessName::OpenClaw,
        display_name: "OpenClaw",
        format: Format::Unsupported,
        servers_key: None,
        path_kind: PathKind::Unsupported,
        supports_mcp_install: false,
        instruction_adapter: InstructionAdapter::None,
        skill_target: SkillTarget::OpenClaw,
        supports_project: false,
    },
];

#[must_use]
/// Returns all harnesses in canonical registry order.
pub fn all() -> &'static [Harness] {
    &REGISTRY
}

/// Looks up one of the nine registered harnesses.
///
/// # Errors
/// Returns `HarnessError::UnknownHarness` for an unregistered name.
pub fn lookup(name: &str) -> Result<&'static Harness, HarnessError> {
    REGISTRY
        .iter()
        .find(|h| h.name.as_str() == name)
        .ok_or_else(|| HarnessError::UnknownHarness(name.to_owned()))
}
#[must_use]
/// Returns all stable harness names in lexical order.
pub fn names() -> Vec<String> {
    let mut result = REGISTRY
        .iter()
        .map(|h| h.name.to_string())
        .collect::<Vec<_>>();
    result.sort();
    result
}
